Create quarkus hal extension for resteasy and reactive

Sgitario opened the pull request:
The quarkus-hal is a new extension that supports both resteasy and resteasy reactive.

- New extension quarkus-hal (mostly copied from quarkus rest data panache). The only change I made is that before, in Rest Data Panache, the jsonb and jackson serializers were serializing the entity into hal and also resolving the hal links. Now, the jsonb and jackson serializers are only serializing the entity that should have been previously populated with the hal links (this is done via server filters in resteasy classic and resteasy reactive).
- Documented in Quarkus Restasy Reactive and Quarkus Resteasy guides

This is a follow-up action from https://github.com/quarkusio/quarkus/pull/25217#issuecomment-1112005083

Sgitario commented:
Breaking changes for Resteasy Classic:
- With these changes, the extension `quarkus-resteasy-links` must be explicitly declared as one of your dependencies. (The same thing was already happening in Resteasy Reactive.
This change affects to `quarkus-hibernate-orm-rest-data-panache` and `quarkus-spring-data-rest`.

geoand commented:
Thanks a lot @Sgitario.

I'll have a look next week,

Sgitario commented:
A possible solution to not add breaking changes in Resteasy Classic is to not remove the auto-generated hal resources in `quarkus-hibernate-rest-data-panache`. Therefore, this dependency would only use the hal API from `quarkus-hal` and nothing else. The disadvantage of this is that we would be still duplicating the resources implementations for JSON and hal, and duplicating the hal JSON serializers in `quarkus-hal` and in `quarkus-hibernate-rest-data-panache` (as the serializers in `quarkus-hibernate-rest-data-panache` are also gathering the HAL links).

Sgitario commented:
> A possible solution to not add breaking changes in Resteasy Classic is to not remove the auto-generated hal resources in `quarkus-hibernate-rest-data-panache`. Therefore, this dependency would only use the hal API from `quarkus-hal` and nothing else. The disadvantage of this is that we would be still duplicating the resources implementations for JSON and hal, and duplicating the hal JSON serializers in `quarkus-hal` and in `quarkus-hibernate-rest-data-panache` (as the serializers in `quarkus-hibernate-rest-data-panache` are also gathering the HAL links).

And another solution could be to change `quarkus-hibernate-orm-rest-data-panache` to use `quarkus-resteas-reactive` by default, so we could add `quarkus-resteasy-reactive-links`. Of course, we'd need to document how to keep using resteasy classic.

geoand commented:
@Sgitario perhaps the description in https://github.com/quarkusio/quarkus/pull/20874 will give you more insight into why things have been done the way they have

Sgitario commented on extensions/panache/hibernate-orm-rest-data-panache/deployment/pom.xml:
```diff
@@ -40,6 +40,11 @@
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-links-deployment</artifactId>
+            <scope>test</scope>
```
I had to explicitly add this dependency, I guess because we had to use the `-deployment` version. However, this is not necessary to be added when using `quarkus-hibernate-orm-rest-data-panache` (there are tests we verify this).

Sgitario commented on extensions/spring-data-rest/deployment/pom.xml:
```diff
@@ -41,6 +41,11 @@
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-links-deployment</artifactId>
+            <scope>test</scope>
```
I had to explicitly add this dependency, I guess because we had to use the `-deployment` version. However, this is not necessary to be added when using `quarkus-spring-data-rest` (there are tests we verify this).

Sgitario commented:
> @Sgitario perhaps the description in #20874 will give you more insight into why things have been done the way they have

PR updated

geoand commented on extensions/hal/runtime/src/main/java/io/quarkus/hal/HalService.java:
```diff
@@ -0,0 +1,79 @@
+    public HalCollectionWrapper toHalCollectionWrapper(Collection<Object> collection, String collectionName,
+            Class<?> entityClass) {
+        List<HalEntityWrapper> items = collection.stream()
+                .map(this::toHalWrapper)
```
Can we please eliminate the use of lambdas in runtime code?

Sgitario replied:
Will delete it. This is a copy/paste of what we were already using in the panache extension and I didn't want to introduce more changes.

geoand commented on extensions/hal/runtime/src/main/java/io/quarkus/hal/HalService.java:
```diff
@@ -0,0 +1,79 @@
+     */
+    public String getSelfLink(Object entity) {
+        return Optional.ofNullable(getInstanceLinks(entity).get(SELF_REF))
+                .map(HalLink::getHref)
```
With Quarkus aiming to provide very small runtime footprint, this is the kind of thing we usually avoid :)

Sgitario replied:
Changed to not using optional and map.

geoand commented on extensions/resteasy-classic/resteasy-links/deployment/src/main/java/io/quarkus/resteasy/links/deployment/LinksProcessor.java:
```diff
@@ -0,0 +1,31 @@
+            if (!capabilities.isPresent(Capability.RESTEASY_JSON_JSONB)
+                    && !capabilities.isPresent(Capability.RESTEASY_JSON_JACKSON)) {
+                throw new IllegalStateException(
+                        "Cannot generate HAL endpoints without a RESTEasy JSON-B or Jackson capability");
```
Just to make this actionable, can add the names of the extensions the user should add?

Sgitario replied:
Changed

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/runtime/src/main/java/io/quarkus/resteasy/reactive/links/runtime/hal/HalServerResponseFilter.java:
```diff
@@ -0,0 +1,91 @@
+        return acceptMediaType.contains(RestMediaType.APPLICATION_HAL_JSON);
+    }
+
+    private Class<?> findEntityClass() {
```
We can leave addressing this for a following PR, but in RESTEasy Reactive we generally try to avoid performing reflection as much as possible as it has a negative effect on both the execution speed and memory consumption. This data is generally available at build time, so we try to leverage it as much as possible.

Sgitario replied:
I've removed the second part of this method. For the first part, we can do it in the next iteration.

geoand replied:
So the second part was not needed?

Sgitario replied:
No, it wasn't needed at all.

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/main/java/io/quarkus/resteasy/reactive/links/deployment/LinksProcessor.java:
```diff
@@ -75,6 +80,24 @@ AdditionalBeanBuildItem registerRestLinksProviderProducer() {
+            if (!capabilities.isPresent(Capability.RESTEASY_REACTIVE_JSON_JSONB) && !capabilities.isPresent(
+                    Capability.RESTEASY_REACTIVE_JSON_JACKSON)) {
+                throw new IllegalStateException(
+                        "Cannot generate HAL endpoints without a RESTEasy JSON-B or Jackson capability");
```
Same as a previous comment I had added

Sgitario replied:
changed

geoand reviewed (commented):
This is great work, thanks!

I've just added a few comments

geoand commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/test/java/io/quarkus/resteasy/reactive/links/deployment/AbstractHalLinksTest.java:
```diff
@@ -0,0 +1,38 @@
+
+public abstract class AbstractHalLinksTest {
+    @BeforeAll
+    public static void setup() {
```
Should this be restored?

Sgitario replied:
I'm not sure about what you mean here

geoand replied:
`RestAssured.defaultParser` is being set before all tests are run. It's previous value should probably be restored or we could potentially encounter hard to debug issues in other tests

Sgitario replied:
I've just configured that setting the defaultParser was not necessary either :/ 
Deleted from the resteasy classic links and the reactive links.

Sgitario committed:
Create quarkus hal extension for resteasy and reactive

The quarkus-hal is a new extension that supports both resteasy and resteasy reactive.
